virtcontainers/qemu_arm64.go

```diff
@@ -45,6 +48,65 @@ var supportedQemuMachines = []govmmQemu.Machine{
 	},
 }
 
+// Logger returns a logrus logger appropriate for logging qemu-aarch64 messages
+func qemuArmLogger() *logrus.Entry {
+	return virtLog.WithField("subsystem", "qemu-aarch64")
+}
+
+// On ARM platform, we have different GIC interrupt controllers. Different
+// GIC supports different QEMU parameters for virtual GIC and max VCPUs
+var hostGICVersion = getHostGICVersion()
+
+// We will access this file on host to detect host GIC version
+var gicProfile = "/proc/interrupts"
+
+// Detect the host GIC version.
+// Success: return the number of GIC version
+// Failed: return 0
+func getHostGICVersion() (version uint32) {
+	bytes, err := ioutil.ReadFile(gicProfile)
+	if err != nil {
+		qemuArmLogger().WithField("GIC profile", gicProfile).WithError(err).Error("Failed to parse GIC profile")
+	}
+
+	if strings.Contains(string(bytes), "GICv2") {
+		return 2
+	}
+
+	if strings.Contains(string(bytes), "GICv3") {
+		return 3
+	}
+
+	if strings.Contains(string(bytes), "GICv4") {
+		return 4
+	}
+
+	return 0
+}
+
+// QEMU supports GICv2, GICv3 and host parameters for gic-version. The host
+// parameter will let QEMU detect GIC version by itself. This parameter
+// will work properly when host GIC version is GICv2 or GICv3. But the
+// detection will failed when host GIC is gicv4 or higher. In this case,
+// we have to detect the host GIC version manually and force QEMU to use
+// GICv3 when host GIC is GICv4 or higher.
+var guestGICversion = getGuestGICVersion()
+
+func getGuestGICVersion() (version string) {
```

Comment:
Nit: For parity with `getHostGICVersion()`, it might be better if `getGuestGICVersion()` simply returned `2`, `3` or `host`. That would make `defaultQemuMachineOptions` a little clearer as it could become:

```
var defaultQemuMachineOptions = "usb=off,accel=kvm," + "gic-version=" + guestGICversion
```

or maybe just:

```
var defaultQemuMachineOptions = "usb=off,accel=kvm," + "gic-version=" + getGuestGICVersion()
```